crates/sui-light-client/src/proof/ocs.rs

```diff
@@ -0,0 +1,258 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use fastcrypto::hash::Blake2b256;
+use fastcrypto::merkle::{MerkleNonInclusionProof, MerkleProof, MerkleTree, Node};
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use sui_types::digests::ObjectDigest;
+use sui_types::messages_checkpoint::{
+    CheckpointArtifacts, CheckpointArtifactsDigest, ObjectCheckpointState,
+};
+use sui_types::{
+    base_types::ObjectID, full_checkpoint_content::CheckpointData,
+    messages_checkpoint::VerifiedCheckpoint,
+};
+
+use crate::proof::{
+    base::{Proof, ProofBuilder, ProofContents, ProofContentsVerifier, ProofTarget},
+    error::{ProofError, ProofResult},
+};
+
+/// A target for a proof about the state of an object w.r.t a checkpoint.
+/// OCS stands for ObjectCheckpointState
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct OCSTarget {
+    pub id: ObjectID,
+    pub digest: Option<ObjectDigest>,
+    pub target_type: OCSTargetType,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub enum OCSTargetType {
+    /// A proof that state of an object o is d at checkpoint c. For cases where o is updated in c.
+    Inclusion,
+    /// A proof that o was not updated in c. For cases where o is not updated in c.
+    NonInclusion,
+}
+
+impl OCSTarget {
+    pub fn new(
+        id: ObjectID,
+        digest: Option<ObjectDigest>,
+        proof_type: OCSTargetType,
+    ) -> ProofResult<Self> {
```

Comment:
Is there any need to have the generic `new` method? I like the intent-specific methods below, which don't need an error return (or a proof_type argument)

Reply:
fair, i removed it. it is kind of unnecessary.